daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/PackedBinaryTraits.scala

```diff
@@ -69,17 +77,30 @@ abstract class PackedBinaryDecimalBaseParser(
 }
 
 abstract class PackedBinaryIntegerBaseParser(
-  override val context: ElementRuntimeData,
-  signed: Boolean = false
+  override val context: ElementRuntimeData
 ) extends PrimParser
   with PackedBinaryConversion {
   override lazy val runtimeDependencies = Vector()
 
+  val signed = {
+    context.optPrimType.get match {
+      case n: NodeInfo.PrimType.PrimNumeric => n.isSigned
+      // context.optPrimType can be of type date/time via ConvertZonedCombinator
+      case _ => false
+    }
+  }
   protected def getBitLength(s: ParseOrUnparseState): Int
 
   def parse(start: PState): Unit = {
     val nBits = getBitLength(start)
-    if (nBits == 0) return // zero length is used for outputValueCalc often.
+    if (nBits == 0) {
```

Comment:
I can't find where we check the packed length is a multiple of 4 bits. Unless you're referring to what we do in the function below?

```scala
def bcdFromBigIntegerLength(absBigIntAsString: String, minLengthInBits: Int): (Int, Int) = {
    val numDigits = absBigIntAsString.length
    // Need to have an even number of digits to fill out a complete byte
    val requiredBitLen = if (numDigits % 2 == 0) (numDigits * 4) else ((numDigits + 1) * 4)
    val bitLen = scala.math.max(minLengthInBits, requiredBitLen)
    val numBytes = (bitLen / 8)
    val leadingZeros = bitLen / 4 - numDigits
    (numBytes, leadingZeros)
  }
```

Reply:
Op, nvm. I found it in org.apache.daffodil.core.grammar.ElementBaseGrammarMixin

```scala
 if ((binaryNumberKnownLengthInBits != -1) && (binaryNumberKnownLengthInBits % 4) != 0)
```

Reply:
This logic doesn't preclude nBits == 0, and it only checks for when nBits is known. I think we should leave in the 0 bits check as it is now, because it covers when the nBits is known and when it is calculated. Thoughts?

Reply:
Yeah, you're right, we can't rely entirely compile time validation since lengths could be runtime calculate.

In fact, in addition to checking it's not zero length, we also need to check it's a multiple for 4 at runtime. Right now during runtime, we just ask for a byte array that is nbits long and the packed logic assumes the array is completely full, even though it could have partial fragment bytes.

So I think we do need to keep this 0 check, but also need to add a % 4 == 0 check.